draft-ietf-tls-esni.md

```diff
@@ -488,32 +478,56 @@ by the server. (Recall that servers might provide multiple ESNIRecord values in
 to a ESNI record query, each containing an ESNIKeys value.) The client MUST NOT send encrypted SNI using groups or
 cipher suites not advertised by the server.
 
-When offering an encrypted SNI, the client MUST NOT offer to resume any non-ESNI
-PSKs. It additionally MUST NOT offer to resume any sessions for TLS 1.2 or
-below.
+When offering an encrypted SNI, the client MUST NOT offer a session ticket
+obtained through a non-ESNI TLS session. It additionally MUST NOT offer to
+resume any sessions for TLS 1.2 or below.
+
+### Generating ClientHello {#generate-clienthello}
+
+The client then builds the first ClientHello, that has the following properties:
+
+* Contains a "server_name" extension that carries the true SNI DNS name in
+  cleartext.
+* Immediately following the "server_name" extension, contains an
+  "encrypted_server_name" extension carrying the ClientEncryptedSNI structure of
+  type `esni_parameters`.
+
+The client MUST NOT send a "cached_info" extension {{!RFC7924}} with a
+CachedObject entry whose CachedInformationType is "cert".
+
+A ClientHello generated in response to a HelloRetryRequest MUST contain a
+"server_name" extension carrying the public name.  This message MUST NOT contain
+an "encrypted_server_name" extension.
+
+These ClientHellos are fed into the Transcript Hash ({{RFC8446}}; Section
+4.4.1).
+
+### ESNI Transformation {#esni-transformation}
+
+Then, before submitting the first ClientHello to the record protocol, the client
+transforms the ClientHello by replacing the aforementioned two extensions with
+the following pair:
+
+* A "server_name" extension that contains ESNIKeys.public_name.
+* Immediately following the "server_name" extension, an "encrypted_server_name"
+  extension carrying the ClientEncryptedSNI structure of type `esni_encrypted`.
+
```

Comment:
See the issue raised by Stephen Farrell. Sending the public name is a nice default, but it should be possible to replace that by a client chosen value.

Reply:
While I appreciate the possibility, I think that's an open issue orthogonal to this PR.